src/data/fintype/basic.lean

```diff
@@ -13,11 +11,58 @@ import data.array.lemmas
 import order.well_founded
 import group_theory.perm.basic
 
+/-!
+# Finite types
+
+This file defines a typeclass to state that a type is finite.
+
+## Main declarations
+
+* `fintype α`:  Typeclass saying that a type is finite. It takes as fields a `finset` and a proof
+  that all terms of type `α` are in it.
+* `finset.univ`: The finset of all elements of a fintype.
+* `fintype.card α`: Cardinality of a fintype. Also `univ.card`.
+* `perms_of_finset s`: The finset of permutations of the finset `s`.
+* `fintype.trunc_equiv_of_card_eq` `fintype.equiv_of_card_eq`
+* `fintype.trunc_equiv_fin`: A fintype `α` is computably equivalent to `fin (card α)`. The
+  `trunc`-free, noncomputable version is `fintype.equiv_fin`.
+* `fin.equiv_iff_eq`: `fin m ≃ fin n` iff `m = n`.
+* `infinite α`: Typeclass saying that a type is infinite. Defined as `fintype α → false`.
+* `not_fintype`: No `fintype` is infinite.
+* `infinite.nat_embedding`: An embedding of `ℕ` into an infinite type.
+
+We also provide the following versions of the pigeonholes principle.
+* `fintype.exists_ne_map_eq_of_card_lt` and `is_empty_of_card_lt`: Finitely many pigeons and
+  pigeonholes. Weak formulation.
+* `fintype.exists_ne_map_eq_of_infinite`: Infinitely many pigeons in finitely many pigeonholes.
+  Weak formulation.
+* `fintype.exists_infinite_fiber`: Infinitely many pigeons in finitely many pigeonholes. Strong
+  formulation.
+Some more is to be found in `data.fintype.card_embedding`.
+
+## Instances
+
+Among others, we provide `fintype` instances for
+* A `subtype` of a fintype. See `fintype.subtype`.
+* The `option` of a fintype.
+* The product of two fintypes.
+* The sum of two fintypes.
+* `Prop`.
+* Types which have a surjection from/an injection to a fintype. See `fintype.of_injective` and
+  `fintype.of_surjective`.
```

Comment:
This isn't an instance - this is just machinery to construct instances with. I think this belongs in a separate list.

Reply:
What about now? I'm not very inspired.